qiskit_experiments/library/characterization/analysis/multi_state_discrimination_analysis.py

```diff
@@ -0,0 +1,147 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Multi state discrimination analysis."""
+
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+
+from typing import List, Tuple
+
+from qiskit.providers.options import Options
+from qiskit_experiments.framework import BaseAnalysis, AnalysisResultData, ExperimentData
+from qiskit_experiments.framework.matplotlib import get_non_gui_ax
+from qiskit_experiments.data_processing import SkCLF
+
+from sklearn.inspection import DecisionBoundaryDisplay
+from sklearn.linear_model import SGDClassifier
+
+
+class MultiStateDiscriminationAnalysis(BaseAnalysis):
+    """This class will fit the discriminator to the data."""
+
+    @classmethod
+    def _default_options(cls) -> Options:
+        """Return default analysis options.
+
+        Analysis Options:
+            plot (bool): Set ``True`` to create figure for fit result.
+            ax(AxesSubplot): Optional. A matplotlib axis object to draw.
+            discriminator: The discriminator to classify the data.
+        """
+        options = super()._default_options()
+        options.plot = True
+        options.ax = None
+        options.discriminator = SkCLF(SGDClassifier(loss="modified_huber", max_iter=1000,
+                                                    tol=1e-3))
+        return options
+
+    def _run_analysis(
+            self,
+            experiment_data: ExperimentData,
+    ) -> Tuple[List[AnalysisResultData], List["matplotlib.figure.Figure"]]:
+        """Train a discriminator based on the experiment data.
+        Args:
+            experiment_data: the data obtained from the experiment
+        Returns:
+            The configuration of the trained discriminator and the IQ plot.
+        """
+
+        # number of states and shots
+        n_states = len(experiment_data.data())
+        NUM_SHOTS = len(experiment_data.data()[0]['memory'])
+
+        # Process the data and get labels
+        result_data, fit_state = [], []
+        for i in range(n_states):
+            state_data = []
+            for j in range(NUM_SHOTS):
+                re, im = experiment_data.data()[i]['memory'][j][0]
+                state_data.append(re + 1j * im)
+                fit_state.append(experiment_data.data()[i]['metadata']['label'])
+            result_data.append(np.array(state_data))
```

Comment:
I would have kept real and imaginary separate in a list of length two. What is the advantage of doing things this way, i.e. with complex values? Also, afterwards you split out the data into a list where each entry is a list of length two with real and imaginary. Looks like you don't need this escaped into complex numbers.

Reply:
Addressed in [c4843e9](https://github.com/Qiskit/qiskit-experiments/pull/936/commits/c4843e97ef5698a5400f02657c41fb6b39e8d655). Removed all the complex numbers logic including the reshape vector function.